### PR TITLE
Add Slack Directory Loader

### PR DESCRIPTION
**Pull request by vowelparrot**
Fixes linting issue from #2835 

Adds a loader for Slack Exports which can be a very valuable source of
knowledge to use for internal QA bots and other use cases.

```py
# Export data from your Slack Workspace first.
from langchain.document_loaders import SLackDirectoryLoader

SLACK_WORKSPACE_URL = "https://awesome.slack.com"

loader = ("Slack_Exports", SLACK_WORKSPACE_URL)
docs = loader.load()
```

### CODE REVIEW

File: langchain/document_loaders/slack_directory.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+        Including the URL will turn sources into links
+        """
+        self.file_path = path
+        self.workspace_url = workspace_url
+        self.channel_id_map = self._get_channel_id_map()
+
+    def _get_channel_id_map(self) -> Dict[str, str]:
+        """Get a dictionary mapping channel names to their respective IDs."""
+        channels_json_path = Path(self.file_path) / "channels.json"
```
Comment (dev2049) [inline langchain/document_loaders/slack_directory.py:23]: should we just store as Path in init

File: langchain/document_loaders/slack_directory.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                channels = json.load(f)
+            return {channel["name"]: channel["id"] for channel in channels}
+        return {}
+
+    def load(self) -> List[Document]:
+        """Load documents."""
+        channel_paths = list(Path(self.file_path).glob("*"))
+        docs = []
+        for channel_path in channel_paths:
+            if channel_path.is_dir():
```
Comment (dev2049) [inline langchain/document_loaders/slack_directory.py:35]: in name of not nesting too much, could do 
```
if not channel_path.is_dir():
  continue
...
```

File: langchain/document_loaders/slack_directory.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                channel_name = channel_path.name
+                json_files = list(channel_path.glob("*.json"))
+                for json_file in json_files:
+                    with open(json_file, encoding="utf-8") as f:
+                        messages = json.load(f)
+                    for message in messages:
+                        text = message.get("text", "")
```
Comment (dev2049) [inline langchain/document_loaders/slack_directory.py:42]: maybe L42-L60 justifies a helper fn, something like `msg_json_to_doc`?

**Review by dev2049 — COMMENTED** [review]
worth adding simple test or two?

**Comment by vowelparrot** [comment]
@dev2049 refactored, added a couple tests as well.

I also changed the interface to take in a path of a zip file. Why? The user otherwise has to manually unzip the folder, and the default nesting of the export is flat, meaning the directories will all be mixed in if the user unzips it in a pre-populated directory.

This way the user can simply download a zip file and point the loader there and we wont try to load spurious `json` files

**Review by dev2049 — APPROVED** [review]

**Pull request merged**